### DISCUSSION TITLE
Can I print a panic to the main screen while in the alternate screen in termwiz?

### DISCUSSION THREAD

**Discussion by groves** [wezterm/wezterm#2749] [Q&A] [answered]
I have a termwiz app that's using the alternate screen. UnixTerminal::drop exits the alternate screen, which is great for almost everything.

My one issue is that if my code panics, [Rust's default panic behavior](https://doc.rust-lang.org/std/macro.panic.html) is to print the panic payload to stderr and then unwind. That means if I panic while in the alternate screen, that stderr print shows up on the alternate screen, Rust unwinds and drops the Terminal which exits the alternate screen, and my app exits without any visible error text. Even if I [set my own panic hook](https://doc.rust-lang.org/std/panic/fn.set_hook.html), I don't see a great way to get out of the alternate screen as the function passed to that needs to be non-mutable and exit_alternate_screen is mutable.

Have other termwiz users found a good way to handle this? I'm thinking I could create a hook that logs the panic message to a file and sets a global boolean. Then I could check for that global boolean in my app's drop code and print to check the log to stderr there as I would've exited the alternate screen at that point. Feels kinda gross though, so I'm hoping someone has a cleaner way.

**Comment by groves** [comment, marked as answer]
I'm exiting alternate screen and printing manually for now:

```rust
   panic::set_hook(Box::new(|info| {
        let _ = write!(
            stderr(),
            "{}",
            CSI::Mode(Mode::ResetDecPrivateMode(DecPrivateMode::Code(
                DecPrivateModeCode::ClearAndEnableAlternateScreen
            ))),
        );
        let thread = thread::current();
        let thread = thread.name().unwrap_or("<unnamed>");
        let msg = match info.payload().downcast_ref::<&'static str>() {
            Some(s) => *s,
            None => match info.payload().downcast_ref::<String>() {
                Some(s) => &**s,
                None => "Box<Any>",
            },
        };
        let _ = match info.location() {
            Some(location) => {
                write!(
                    stderr(),
                    "thread '{}' panicked at '{}': {}:{}\r\n",
                    thread,
                    msg,
                    location.file(),
                    location.line(),
                )
            }
            None => {
                write!(stderr(), "thread '{}' panicked at '{}'\r\n", thread, msg,)
            }
        };
    }));
```

I can't use the default hook as this still leaves the terminal in raw mode at the point when it's called, and it seems even worse for me to exit raw mode here. This is at least a touch simpler than the boolean code I was proposing above, but it's still quite gross.

**Reply by wez** [reply]
If the panic somehow results in the main flow of your program unwinding and allowing Drop impls to run, a possible approach might be to add a wrapper around the `Terminal`; here's a sketch of what I mean. I think you'll need to tweak it a little to get it to actually work, but hopefully it demonstrates the idea:

```rust
static PANIC_MESSAGE: Mutex<Option<String>> = Mutex::new(None);

struct Wrap {
  terminal: SystemTerminal,
}

impl Drop for Wrap {
  fn drop(&mut self) {
    drop(self.terminal);
    if let Some(message) = PANIC_MESSAGE.lock().unwrap().take() {
      eprintln!("PANIC: {message}");
    }
  }
}
```

then in your panic hook, set `PANIC_MESSAGE`.

**Reply by groves** [reply]
Thanks! That does feel nicer than trying to exit alternate screen myself.

I don't think it's possible to do something like `drop(self.terminal);` or at least I couldn't figure out how to do it. drop takes ownership of the value, and we can't transfer ownership out of Wrap and keep it valid.

I ended up writing this:

```rust
static PANIC_MESSAGE: Mutex<Option<String>> = Mutex::new(None);

struct DropLast {}
impl Drop for DropLast {
    fn drop(&mut self) {
        if let Some(message) = PANIC_MESSAGE.lock().unwrap().take() {
            println!("PANIC: {message}");
        }
    }
}

struct Wrap {
    term: BufferedTerminal<SystemTerminal>,
    _dl: DropLast,
}
```

That takes advantage of the fact that drops are guaranteed to be called in the same order as field declaration. drop on DropLast should always be called after term.drop has been called to exit the alternate screen.

Then I setup the PANIC_MESSAGE setting hook like this:

```rust
    panic::set_hook(Box::new(|info| {
        let thread = thread::current();
        let thread = thread.name().unwrap_or("<unnamed>");
        let msg = match info.payload().downcast_ref::<&'static str>() {
            Some(s) => *s,
            None => match info.payload().downcast_ref::<String>() {
                Some(s) => &**s,
                None => "Box<Any>",
            },
        };
        let mut pmsg = PANIC_MESSAGE.lock().unwrap();
        *pmsg = Some(match info.location() {
            Some(location) => {
                format!(
                    "thread '{}' panicked at '{}': {}:{}\r\n",
                    thread,
                    msg,
                    location.file(),
                    location.line(),
                )
            }
            None => {
                format!("thread '{}' panicked at '{}'\r\n", thread, msg,)
            }
        });
    }));
```

Seems to work. Thanks again for the help!
